src/aoscene.cpp

```diff
@@ -1,119 +1,374 @@
-  }
-  else {
-    QPixmap background(background_path);
-    auto transform_mode = Qt::FastTransformation;
-    if (background.height() > f_h) // We are downscaling, use anti-aliasing.
-      transform_mode = Qt::SmoothTransformation;
-
-    background = background.scaledToHeight(f_h, transform_mode);
-    this->resize(background.size());
-    this->setPixmap(background);
-    QLabel::move(
-        x + (f_w - background.width()) / 2,
-        y + (f_h - background.height()) /
-                2); // Always center horizontally, always center vertically
   }
-  last_image = background_path;
 }
 
-void AOScene::set_legacy_desk(QString p_image)
+void AOScene::load_network_effects()
 {
+  movie_effects.clear();
+  movie_effects.resize(max_frames);
+  // Order is important!!!
+  QStringList effects_list = {"shake", "flash", "sfx^"};
 
-  QString desk_path =
-      ao_app->get_image_suffix(ao_app->get_background_path(p_image));
-  if (!file_exists(desk_path)) // If image is missing, clear current image
-  {
-    this->clear();
-    this->setMovie(nullptr);
+  // Determines which list is smaller - effects_list or network_strings - and
+  // uses it as basis for the loop. This way, incomplete network_strings would
+  // still be parsed, and excess/unaccounted for networked information is
+  // omitted.
+  int effects_size = qMin(effects_list.size(), network_strings.size());
 
-    m_movie->stop();
-    last_image = "";
-    return;
+  for (int i = 0; i < effects_size; ++i) {
+    QString netstring = network_strings.at(i);
+    QStringList emote_splits = netstring.split("^");
+    foreach (QString emote, emote_splits) {
+      QStringList parsed = emote.split("|");
+      if (parsed.size() <= 0 || parsed.at(0) != m_emote)
+        continue;
+      foreach (QString frame_data, parsed) {
+        QStringList frame_split = frame_data.split("=");
+        if (frame_split.size() <=
+            1) // We might still be hanging at the emote itself (entry 0).
+          continue;
+        int f_frame = frame_split.at(0).toInt();
+        if (f_frame >= max_frames) {
+          qDebug() << "Warning: out of bounds" << effects_list[i] << "frame"
+                   << f_frame << "out of" << max_frames << "for" << m_char
+                   << m_emote;
+          continue;
+        }
+        QString f_data = frame_split.at(1);
+        if (f_data != "") {
+          QString effect = effects_list[i];
+          if (effect == "sfx^") // Currently the only frame result that feeds us
+                                // data, let's yank it in.
+            effect += f_data;
+          qDebug() << effect << f_data << "frame" << f_frame << "for" << m_char
+                   << m_emote;
+          movie_effects[f_frame].append(effect);
+        }
+      }
+    }
   }
+}
 
-  if (file_exists(desk_path) && desk_path == last_image)
-    return;
+void AOScene::play()
+{
+  play_frame_effect(frame);
+  if (max_frames <= 1) {
+    if (play_once)
+      ticker->start(60);
```

Comment:
what is `60`

Reply:
i have no idea, this is verbatim from charmovie

Reply:
uses tick_ms now